kernel/arch/dreamcast/fs/fs_vmu.c

```diff
@@ -251,36 +261,71 @@ static vmu_fh_t *vmu_open_file(maple_device_t * dev, const char *path, int mode)
        then we need to read the old file if there is one. */
     realmode = mode & O_MODE_MASK;
 
+    if(realmode == O_RDONLY && (mode & O_TRUNC)) {
+        errno = EINVAL;
+        goto error;
+    }
+    //if(!(mode & O_META) && strcmp(fd->name, "ICONDATA_VMS") == 0) {
+    //    dbglog(DBG_SOURCE(VMUFS_DEBUG), "vmu_open_file: attempt to open ICONDATA_VMS without raw mode\n");
+    //}
+
     if(realmode == O_RDONLY || ((realmode == O_RDWR || realmode == O_WRONLY) && !(mode & O_TRUNC))) {
         /* Try to open it */
-        rv = vmufs_read(dev, fd->name, &data, &datasize);
+        rv = vmufs_read(dev, fd->name, &data, &datasize, &filetype);
 
-        if(rv < 0) {
-            if(realmode == O_RDWR || realmode == O_WRONLY) {
-                /* In some modes failure is ok -- flag to setup a blank first block. */
-                datasize = -1;
-            }
-            else {
-                free(fd);
-                return NULL;
+        if(rv != 0 && rv != -2) {
+            /* Invalid device, read error or truncated file */
+            errno = EIO;
+            goto error;
+        }
+        else if(rv == -2) {
+            if(realmode == O_RDONLY) {
+                errno = ENOENT;
+                goto error;
             }
+
+            //if(!(mode & O_CREAT)) {
+            //    //dbglog(DBG_INFO, "vmu_open_file: file %s not found\n", path);
+            //    errno = ENOENT;
+            //    goto error;
+            //}
+
+            /* File not found, flag to setup new blocks. */
+            datasize = -1;
+        }
+        else if(filetype != 0x33) {
+            dbglog(DBG_WARNING, "VMUFS: file %s isn't DATA type\n", path);
+            errno = EFTYPE;
+            goto error;
         }
     }
     else {
-        /* We're writing with truncate... flag to setup a blank first block. */
+        /* We're writing with truncate... flag to setup new blocks.*/
+        /* File type is unknown (nothing was readed) */
         datasize = -1;
     }
 
-    /* We were flagged to set up a blank first block */
+    /* We were flagged to set up new blocks */
     if(datasize == -1) {
         data = malloc(512);
         if(data == NULL) {
-            free(fd);
-            return NULL;
+            errno = ENOMEM;
+            goto error;
         }
         datasize = 512;
         memset(data, 0, 512);
-    } else if(!fd->raw && !vmu_pkg_parse(data, datasize, &vmu_pkg)) {
+    } else if(!fd->raw) {
+        /* Parse header but ignore any bad checksum */
+        rv = vmu_pkg_parse(data, datasize, &vmu_pkg);
+        if (rv == -1) {
+            /* Invalid checksum */
+            errno = EILSEQ;
+            goto error;
+        } else if(rv == -2) {
+            /* vmufs_read() doesn't lie about the file size, file is totally corrupted */
+            errno = EIO;
+            goto error;
+        }
```

Comment:
That is wrong, sorry. `vmu_open_file()` should be able to still open files that have no header. I made sure it was possible when I added support for transparent headers, so that it wouldn't unnecessarily break ABI with old code.